File: SampleApps/ObjC/SampleAppWakeUp/SampleAppWakeUp/Services/LocationService/MistService.m
//
//  MistService.m
//  SampleAppWakeUp
//
//  Created by Akhil Rana on 24/11/23.
//

@import Foundation;
@import MistSDK;
#import "MistService.h"
#import "SampleAppWakeUp-Swift.h"

@interface MistService () <MistSDKManagerDelegate>
@property (nonatomic, strong, readwrite, nullable) MistMap *currentMap;
@property (nonatomic, strong) MistSDKManager *sdkManager;
@end

@implementation MistService

- (instancetype)initWithToken:(NSString *)token orgId:(NSString *)orgId {
    self = [super init];
    if (self) {
        _sdkManager = [[MistSDKManager alloc] initWithToken:token orgId:orgId];
        _sdkManager.delegate = self;
    }
    return self;
}

- (void)start {
    [self.sdkManager start];
    self.isMistSDKStarted = YES;
}

- (void)stop {
    [self.sdkManager stop];
    self.isMistSDKStarted = NO;
}

// MARK: - MistSDKManagerDelegate

- (void)didUpdateMap:(MistMap *)map {
    self.currentMap = map;
    NSURL *url = [NSURL URLWithString:map.url];
    [self.delegate didUpdateMap:url];
}

- (void)didUpdateLocation:(CGPoint)location {
    if (self.currentMap) {
        CGFloat scaledX = location.x / self.currentMap.ppm;
        CGFloat scaledY = location.y / self.currentMap.ppm;
        CGPoint clientLocation = CGPointMake(scaledX, scaledY);
        [self.delegate didUpdateLocation:clientLocation];
    }
}

- (void)didFailWithError:(NSString *)error {
    [self.delegate didFailWithError:error];
}

@end

File: SampleApps/Swift/~2.0.8/SampleBlueDotExperience/SampleBlueDotExperience/Services/MistService.swift
//
//  MistService.swift
//  SampleBlueDotExperience
//
//  Created by Rajesh Vishwakarma on 27/01/23.
//

import Foundation
import MistSDK

protocol MistService {
    func start()
    func stop()
}

protocol MistServiceDelegate: AnyObject {
    func didUpdateMap(_ map: URL)
    func didUpdateLocation(_ location: CGPoint)
    func didFailed(with error: String?)
}

class RealMistService: MistService {
    private let mistManager: IndoorLocationManager
    private var currentMap: MistMap?

    weak var delegate: MistServiceDelegate?
    
    init(token: String, delegate: MistServiceDelegate) {
        self.delegate = delegate
        self.mistManager = IndoorLocationManager.sharedInstance(token)
    }
    
    func start() {
        mistManager.startWithIndoorLocationDelegate(self)
    }
    
    func stop() {
        mistManager.stop()
    }
}

// MARK :- IndoorLocationDelegate
extension RealMistService: IndoorLocationDelegate {
    
    func didUpdateMap(_ map: MistMap!) {
        currentMap = map
        guard let mapURL = URL(string: map.url) else { return }
        delegate?.didUpdateMap(mapURL)
        debugPrint(">>> didUpdate MistMap mapName=\(map.name.description)")
    }
    
    func didUpdateRelativeLocation(_ relativeLocation: MistPoint!) {
        guard let map = currentMap else { return }
        let location = CGPoint(x: relativeLocation.x * map.ppm, y: relativeLocation.y * map.ppm)
        debugPrint(">>> didUpdateRelativeLocation x = \(location.x) y = \(location.y)")
        delegate?.didUpdateLocation(location)
    }
    
    func didReceivedOrgInfoWithTokenName(_ tokenName: String!, andOrgID orgID: String!) {
        debugPrint(">>> didReceivedOrgInfo tokenName = \(tokenName.description) orgID = \(orgID.description)")
    }
    
    func didErrorOccurWithType(_ errorType: Error, andMessage errorMessage: String!) {
        debugPrint(">>> didErrorOccur errorType = \(errorType) errorMessage = \(String(describing: errorMessage))")
        delegate?.didFailed(with: errorMessage)
    }
}
